### ISSUE TITLE
Reified Types for "multi" Providers.

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/angular#649]
As implemented today, `multi` providers have a problem with Dart 2.0 compatibility:

```dart
class SomeValidator implements Validator { ... }

const validationModule = const [
  const Provider(NG_VALIDATORS, useClass: SomeValidator, multi: true),
];
````

```dart
@Component()
class UseValidators {
  UseValidators(@Inject(NG_VALIDATORS) List<Validator> validators) {
    ...
  }
}
```

In DDC, and upcoming in Dart 2.0, this is a _runtime_ error 😿 - I *think* all `multi` providers are _always_ reified as `List<dynamic>`. [Here is the current implementation in `ReflectiveInjector`](https://github.com/dart-lang/angular/blob/master/angular/lib/src/di/injector/reflective.dart#L144-L152):

```dart
List<dynamic> _resolveWithReflectorMulti(RuntimeProvider<dynamic> provider) {
  final results = <dynamic>[];
  for (final multiProvider in _multiProviders) {
    if (identical(multiProvider.token, provider.token)) {
      results.add(_resolveWithReflector(multiProvider));
    }
  }
  return results;
}
```

... and the [current implementation in `AppView`](https://github.com/dart-lang/angular/blob/master/angular/lib/src/compiler/view_compiler/compile_element.dart#L658-L661):

```dart
if (isMulti) {
  resolvedProviderValueExpr = o.literalArr(providerValueExpressions);
  type = new o.ArrayType(o.DYNAMIC_TYPE);
} else {
  ...
}
```

... not including the in-progress `GeneratedInjector`.

**tl;dr**, in Dart 2.x type system implementations, we always try to use `List<dynamic>`, but user code as often structured to be `List<T>`, where `T` is the type the provider provides. In the above example, we'd expect to get `List<Validator>`.

-----

## Proposal

* Add `MultiProvider<T>` to dependency injection.
* Add a private `_listOf` method to `MultiProvider`.
* Treat `MultiProvider<T>` special in both generated implementations.
* Remove `multi: ...` as an option from AngularDart.

### `MultiProvider<T>`

```dart
abstract class MultiProvider<T> implements Provider<T> {
  const factory MultiProvider(
    Object token,
    Type useClass,
    ... etc ...
  ) = ...

  /// Creates a reified `List<T>`, where `T` is the `MultiProvider`'s expected type.
  List<T> _listOf() => <T>[];
}
```

### Changes to `ReflectiveInjector`

```dart
List<T> _resolveWithReflectorMulti<T>(MultiProvider<T> provider) {
  final /* List<T> */ results = internalMultiListOf(provider);
  for (final multiProvider in _multiProviders) {
    if (identical(multiProvider.token, provider.token)) {
      results.add(_resolveWithReflector(multiProvider));
    }
  }
  return results;
}
```

### Changes to `AppView` (via `compile_element`)

```dart
if (isMulti) {
  resolvedProviderValueExpr = o.literalArr(providerValueExpressions);
  type = new o.ArrayType(providerTypeParameter);
} else {
  ...
}
```

-----

I imagine this would take about a week of work, not including deprecations.

/cc @vsmenon @jmesserly @ferhatb @chalin for input.

**Comment by matanlurey** [comment]
I'm moving ahead with this.

**Comment by matanlurey** [comment]
Landed.
